mission_control/include/mission_control/behaviors/waypoint.h

```diff
@@ -51,32 +51,102 @@
 #ifndef MISSION_CONTROL_BEHAVIORS_WAYPOINT_H
 #define MISSION_CONTROL_BEHAVIORS_WAYPOINT_H
 
+#include <behaviortree_cpp_v3/basic_types.h>
+#include <behaviortree_cpp_v3/behavior_tree.h>
+#include <behaviortree_cpp_v3/bt_factory.h>
+#include <ros/ros.h>
+
+#include "mission_control/Waypoint.h"
 #include "mission_control/behavior.h"
-#include "mission_control/Waypoint.h"  // this is the ROS Message
+#include "pose_estimator/CorrectedData.h"
+
+using namespace BT;
+
+// WGS84 Parameters
+#define WGS84_A 6378137.0         // major axis
+#define WGS84_B 6356752.31424518  // minor axis
+#define WGS84_F 0.0033528107      // ellipsoid flattening
+#define WGS84_E 0.0818191908      // first eccentricity
+#define WGS84_EP 0.0820944379     // second eccentricity
+
+// UTM Parameters
+#define UTM_K0 0.9996                    // scale factor
+#define UTM_FE 500000.0                  // false easting
+#define UTM_FN_N 0.0                     // false northing, northern hemisphere
+#define UTM_FN_S 10000000.0              // false northing, southern hemisphere
+#define UTM_E2 (WGS84_E * WGS84_E)       // e^2
+#define UTM_E4 (UTM_E2 * UTM_E2)         // e^4
+#define UTM_E6 (UTM_E4 * UTM_E2)         // e^6
+#define UTM_EP2 (UTM_E2 / (1 - UTM_E2))  // e'^2
 
 namespace mission_control
 {
+struct WaypointData
+{
+  double depth;
+  double altitude;
+  double latitude;
+  double longitude;
+  double wp_radius;
+  double speed_knots;
+};
 
+}  // namespace mission_control
+
+namespace BT
+{
+template <>
+inline mission_control::WaypointData convertFromString<mission_control::WaypointData>(
+    StringView str)
+
+{
+  // real numbers separated by semicolons
+  auto parts = splitString(str, ';');
+  if (parts.size() != 6)
+  {
+    throw RuntimeError("invalid input)");
+  }
+  else
+  {
+    mission_control::WaypointData output;
+    output.depth = convertFromString<double>(parts[0]);
+    output.altitude = convertFromString<double>(parts[1]);
+    output.latitude = convertFromString<double>(parts[2]);
+    output.longitude = convertFromString<double>(parts[3]);
+    output.wp_radius = convertFromString<double>(parts[4]);
+    output.speed_knots = convertFromString<double>(parts[5]);
+    return output;
+  }
+}
+}  // namespace BT
+
+namespace mission_control
+{
 class WaypointBehavior : public Behavior
 {
  public:
-  WaypointBehavior();
-  virtual ~WaypointBehavior();
+  WaypointBehavior(const std::string& name, const BT::NodeConfiguration& config);
 
-  virtual bool parseMissionFileParams();
-  bool getParams(ros::NodeHandle nh);
+  BT::NodeStatus getBehaviorStatus();
+  void abortBehavior();
+  void publishMsg();
 
-  virtual void publishMsg();
-  bool checkCorrectedData(const pose_estimator::CorrectedData& data);
+  static PortsList providedPorts()
+  {
+    const char* description = "depth;altitude;latitude;longitude;wp_radius;speed_knots";
```

Comment:
@patapufate nit: let the compiler optimize for space:
```suggestion
    constexpr char description[] = "depth;altitude;latitude;longitude;wp_radius;speed_knots";
```

Reply:
Will not be fixed because the comment to use different ports.